contributors/design-proposals/storage/pv-monitoring-proposal.md

```diff
@@ -0,0 +1,128 @@
+
+### Monitoring controller: 
+
+Like PV controller, monitoring controller should check PVs’ health condition periodically and taint them if PVs are unhealthy.
+Health checking implementation should be per plugin. Each volume plugin needs to have its own methods to check its volumes.
+
+#### For in-tree volume plugins(except local storage):
+
+We can add a new volume plugin interface: StatusCheckingVolumePlugin.
+```
+  type StatusCheckingVolumePlugin interface {
+       VolumePlugin
+
+      CheckStatus(spec *Spec) (string, error)
+  }
+```
+And each volume plugin will implement it. The entire monitoring controller workflow is:
+
+* Fill PV cache with initial data from etcd
+* Resync and check volumes status periodically
+* Taint PV if the volume status is abnormal
+* Remove the taints if the volume becomes normal again
+
+#### For out-tree volume plugins(except local storage):
+
+We can create a new controller called MonitorController like ProvisionController, 
+which is responsible for creating informers, watching Node and PV events and calling each plugin’s monitor functions. 
+And each volume plugin will create its own monitor to check its volumes’ status.
+
+#### For local storage:
+
+The local storage PV monitor consists of two parts
+
+* create a daemonset on every node, which is responsible for monitoring local PVs in that specific node, no matter the PVs are created manually or by provisioner;
+* create a monitor controller, which is responsible for watching PVs and Nodes events. PVs may be updated if they are unhealthy and we also need to react to node failure event.
+
+At the first phase, we can support local storage monitoring first.
+
+Take local storage as an example, detailed monitoring implementation may be like this:
+
+```
+ func (mon *minitor) CheckStatus(spec *v1.PersistentVolumeSpec) (string, error) {
+        // check if PV is local storage
+        if pv.Spec.Local == nil { 
+                glog.Infof("PV: %s is not local storage", pv.Name)
+                return ...
+        }
+        
+        // check if PV belongs to this node
+        if checkNodeAffinity(...) {
+                glog.Infof("PV: %s does not belong to this node", pv.Name)
+                return ...
+        }
+
+        // check volume health condition depending on device type
+        ...
+
+}
+```
+If monitor finds that one PV is unhealthy, it will mark the PV by adding annotations including timestamp. 
+The reaction controller then can react to this PV depending on the annotations and timestamp.
+
+### Reaction controller:
+
+Reaction controller will react to the PV update event (tainting PV by monitoring controller). 
+Different kinds of apps should have different reactions.
+
+At the first phase, we just consider the statefulset reaction.
+
+statefulset reaction: check the annotation timestamp, if the PV can recover within the predefined time interval, we will do nothing, 
```

Comment:
How does this controller know which StatefulSets to monitor?

Reply:
by monitoring relatedPVs

Reply:
How do we mark which PVs we should monitor?  For example, not all StatefulSets may want this behavior.  So we need a way to be able to say which StatefulSets/PVs should be handled by the reaction controller.

BTW, @jingxu97 just showed me some cool work related to a Snapshot policy controller using the [metacontroller framework](https://github.com/GoogleCloudPlatform/metacontroller).  A CRD is defined that specifies:
* How often a snapshot should be taken
* Label selector to choose which PVCs should follow this policy.

I think we could do something similar for this controller, and have a CRD to define a policy with:
* What conditions to react to
* Time thresholds before reacting
* What should the action be for each condition
* Label selector for which StatefulSets to monitor